File: backend/engines/dj_request_agent.py
from __future__ import annotations

from dataclasses import dataclass, field
import asyncio
import json
import re


ALLOWED_ACTIONS = {
    "play_now",
    "set_direction_and_play",
    "revise_mode_and_play",
    "soft_confirm_and_play",
    "ask_clarifying_question",
    "negative_feedback",
    "continue_current_mode",
}

PLAYABLE_ACTIONS = {
    "play_now",
    "set_direction_and_play",
    "revise_mode_and_play",
    "soft_confirm_and_play",
    "continue_current_mode",
}


@dataclass
class DJDecision:
    action: str
    understood_intent: str = ""
    music_task: dict = field(default_factory=dict)
    queue_policy: dict = field(default_factory=dict)
    uncertainty: dict = field(default_factory=dict)
    dj_response: dict = field(default_factory=dict)
    memory_update: dict = field(default_factory=dict)
    raw_text: str = ""

    def to_dict(self) -> dict:
        return {
            "action": self.action,
            "understood_intent": self.understood_intent,
            "music_task": self.music_task,
            "queue_policy": self.queue_policy,
            "uncertainty": self.uncertainty,
            "dj_response": self.dj_response,
            "memory_update": self.memory_update,
            "raw_text": self.raw_text,
        }


class DJRequestAgent:
    def __init__(self, llm, llm_timeout_s: float = 8.0):
        self.llm = llm
        self.llm_timeout_s = llm_timeout_s

    async def decide(self, user_message: str, context_pack: dict | None = None) -> DJDecision:
        clean = self._clean_user_message(user_message)
        if not clean:
            return self._safe_question(clean)
        prompt = self._prompt(clean, context_pack or {})
        try:
            response = await asyncio.wait_for(
                self.llm.chat(
                    prompt,
                    max_tokens=900,
                    system=(
                        "You are FREQME's private AI DJ request agent. "
                        "Understand the user's music intent from context and output only valid JSON."
                    ),
                ),
                timeout=self.llm_timeout_s,
            )
            data = self._parse_json(response)
        except Exception:
            data = {}
        return self._decision_from_data(clean, data)

    def _prompt(self, user_message: str, context_pack: dict) -> str:
        return f"""User just spoke to the AI radio DJ:
{user_message}

Context pack JSON:
{json.dumps(context_pack, ensure_ascii=False, indent=2)}

Decide what the user wants musically. Be an AI DJ, not a keyword classifier.

Rules:
- Do not search the literal user sentence unless the sentence itself is a canonical title.
- Infer fuzzy artist, band, composer, performer, work-family, scene, style, mood, correction, rejection, or continuation intent.
- Most requests should play now or set a direction and play.
- Ask only when ambiguity is high and context cannot resolve it.
- A request can set a multi-track queue policy.
- Memory update should describe user/session preference, not encyclopedia aliases.
- Never expose profile, algorithm, or system wording in the DJ response.

Return only JSON:
{{
  "action": "play_now | set_direction_and_play | revise_mode_and_play | soft_confirm_and_play | ask_clarifying_question | negative_feedback | continue_current_mode",
  "understood_intent": "internal interpretation",
  "music_task": {{
    "type": "specific_track | artist_direction | artist_work_direction | scene_genre_direction | negative_feedback | continuation",
    "primary_entities": [{{"role": "artist|performer|composer|work|genre", "name": "canonical or inferred name"}}],
    "work_hint": "",
    "style_hint": "",
    "search_goals": ["concrete NetEase-friendly search query"],
    "must_not_search_literal_user_sentence": true
  }},
  "queue_policy": {{"duration_tracks": 1, "continue_direction": false, "avoid_repetition": true}},
  "uncertainty": {{"level": "low|medium|high", "reason": "", "should_ask_user": false}},
  "dj_response": {{"speak_now": "short natural Chinese DJ response", "tone": "warm_confident"}},
  "memory_update": {{"session_preference": [], "possible_long_term_preference": [], "negative_constraints": []}}
}}"""

    def _decision_from_data(self, raw_text: str, data: dict) -> DJDecision:
        if not isinstance(data, dict) or not isinstance(data.get("action"), str):
            return self._safe_question(raw_text)
        action = data["action"]
        if action not in ALLOWED_ACTIONS:
            return self._safe_question(raw_text)

        music_task = self._normalize_music_task(data.get("music_task"))
        if action in PLAYABLE_ACTIONS and not self._is_executable_music_task(music_task):
            return self._safe_question(raw_text)

        queue_policy = self._normalize_queue_policy(data.get("queue_policy"), action)
        uncertainty = self._normalize_uncertainty(data.get("uncertainty"), action)
        dj_response = self._normalize_dj_response(data.get("dj_response"))
        memory_update = self._normalize_memory_update(data.get("memory_update"))
        return DJDecision(
            action=action,
            understood_intent=str(data.get("understood_intent") or ""),
            music_task=music_task,
            queue_policy=queue_policy,
            uncertainty=uncertainty,
            dj_response=dj_response,
            memory_update=memory_update,
            raw_text=raw_text,
        )

    def _safe_question(self, raw_text: str) -> DJDecision:
        return DJDecision(
            action="ask_clarifying_question",
            understood_intent="The request is not clear enough to select music safely.",
            music_task={
                "type": "unclear",
                "primary_entities": [],
                "work_hint": "",
                "style_hint": "",
                "search_goals": [],
                "must_not_search_literal_user_sentence": True,
            },
            queue_policy={"duration_tracks": 0, "continue_direction": False, "avoid_repetition": True},
            uncertainty={"level": "high", "reason": "", "should_ask_user": True},
            dj_response={"speak_now": "这个我没接稳，是想听某个歌手，还是这种氛围？", "tone": "warm_clarifying"},
            memory_update={"session_preference": [], "possible_long_term_preference": [], "negative_constraints": []},
            raw_text=raw_text,
        )

    def _clean_user_message(self, user_message: str) -> str:
        clean = " ".join(str(user_message or "").split())[:500]
        if not clean:
            return ""
        return self._strip_accidental_latin_prefix(clean)

    def _strip_accidental_latin_prefix(self, text: str) -> str:
        match = re.match(r"^[A-Za-z]\s*(?=[\u4e00-\u9fff])", text)
        if not match:
            return text
        candidate = text[match.end():].strip()
        if not self._starts_with_request_marker(candidate):
            return text
        return candidate

    def _starts_with_request_marker(self, text: str) -> bool:
        return str(text or "").startswith(
            (
                "我要",
                "我想",
                "想听",
                "想要",
                "来点",
                "放点",
                "播点",
                "播放",
                "点首",
                "点一首",
                "给我",
                "换成",
                "换点",
                "接着",
                "继续",
                "能不能",
                "不能",
                "可以",
            )
        )

    def _parse_json(self, text: str) -> dict:
        raw = str(text or "").strip()
        try:
            return json.loads(raw)
        except Exception:
            pass

        candidates = [match.group(1) for match in re.finditer(r"```(?:json)?\s*(\{.*?\})\s*```", raw, re.S | re.I)]
        candidates.extend(self._json_object_candidates(raw))
        for candidate in candidates:
            try:
                data = json.loads(candidate)
            except Exception:
                continue
            if isinstance(data, dict):
                return data
        return {}

    def _json_object_candidates(self, text: str) -> list[str]:
        candidates = []
        starts = [index for index, char in enumerate(text) if char == "{"]
        for start in starts:
            depth = 0
            in_string = False
            escape = False
            for index in range(start, len(text)):
                char = text[index]
                if in_string:
                    if escape:
                        escape = False
                    elif char == "\\":
                        escape = True
                    elif char == '"':
                        in_string = False
                    continue
                if char == '"':
                    in_string = True
                elif char == "{":
                    depth += 1
                elif char == "}":
                    depth -= 1
                    if depth == 0:
                        candidates.append(text[start : index + 1])
                        break
        return candidates

    def _normalize_music_task(self, value) -> dict:
        source = value if isinstance(value, dict) else {}
        return {
            "type": str(source.get("type") or ""),
            "primary_entities": self._normalize_primary_entities(source.get("primary_entities")),
            "work_hint": str(source.get("work_hint") or ""),
            "style_hint": str(source.get("style_hint") or ""),
            "search_goals": self._normalize_search_goals(source.get("search_goals")),
            "must_not_search_literal_user_sentence": True,
        }

    def _is_executable_music_task(self, music_task: dict) -> bool:
        return any(
            [
                bool(music_task.get("search_goals")),
                bool(music_task.get("primary_entities")),
                bool(str(music_task.get("work_hint") or "").strip()),
                bool(str(music_task.get("style_hint") or "").strip()),
            ]
        )

    def _normalize_search_goals(self, value) -> list[str]:
        if not isinstance(value, list):
            return []
        goals = []
        for item in value:
            goal = str(item or "").strip()
            if goal:
                goals.append(goal)
        return goals

    def _normalize_primary_entities(self, value) -> list[dict]:
        if not isinstance(value, list):
            return []
        entities = []
        for item in value:
            if not isinstance(item, dict):
                continue
            name = str(item.get("name") or "").strip()
            if not name:
                continue
            role = str(item.get("role") or "").strip() or "music_entity"
            entities.append({"role": role, "name": name})
        return entities

    def _normalize_queue_policy(self, value, action: str) -> dict:
        source = value if isinstance(value, dict) else {}
        return {
            "duration_tracks": self._coerce_duration(source.get("duration_tracks"), default=1),
            "continue_direction": self._coerce_bool(
                source.get("continue_direction"),
                default=action in {"set_direction_and_play", "continue_current_mode"},
            ),
            "avoid_repetition": self._coerce_bool(source.get("avoid_repetition"), default=True),
        }

    def _normalize_uncertainty(self, value, action: str) -> dict:
        source = value if isinstance(value, dict) else {}
        level = source.get("level")
        if level not in {"low", "medium", "high"}:
            level = "medium"
        return {
            "level": level,
            "reason": str(source.get("reason") or ""),
            "should_ask_user": self._coerce_bool(
                source.get("should_ask_user"),
                default=action == "ask_clarifying_question",
            ),
        }

    def _normalize_dj_response(self, value) -> dict:
        source = value if isinstance(value, dict) else {}
        return {
            "speak_now": str(source.get("speak_now") or "我先按我理解到的方向接上。"),
            "tone": str(source.get("tone") or "warm_confident"),
        }

    def _normalize_memory_update(self, value) -> dict:
        source = value if isinstance(value, dict) else {}
        return {
            "session_preference": source.get("session_preference") if isinstance(source.get("session_preference"), list) else [],
            "possible_long_term_preference": (
                source.get("possible_long_term_preference")
                if isinstance(source.get("possible_long_term_preference"), list)
                else []
            ),
            "negative_constraints": source.get("negative_constraints") if isinstance(source.get("negative_constraints"), list) else [],
        }

    def _coerce_duration(self, value, default: int) -> int:
        try:
            duration = int(value)
        except (TypeError, ValueError):
            duration = default
        return max(0, min(8, duration))

    def _coerce_bool(self, value, default: bool) -> bool:
        return value if isinstance(value, bool) else default

File: src/radio/djRequestAgent.ts
import type { DJDecision, MemoryPack, MusicTask } from "../types.js";
import type { LLMRouter } from "../services/llmRouter.js";
import { asStringList, compactText, extractJsonObject } from "../utils/text.js";

const allowedActions = new Set<DJDecision["action"]>([
  "play_now",
  "set_direction_and_play",
  "revise_mode_and_play",
  "soft_confirm_and_play",
  "ask_clarifying_question",
  "negative_feedback",
  "continue_current_mode",
]);

const playableActions = new Set<DJDecision["action"]>([
  "play_now",
  "set_direction_and_play",
  "revise_mode_and_play",
  "soft_confirm_and_play",
  "continue_current_mode",
  "negative_feedback",
]);

const validTaskTypes = new Set<MusicTask["type"]>([
  "specific_track",
  "artist_direction",
  "artist_work_direction",
  "scene_genre_direction",
  "continuation",
  "negative_feedback",
  "unclear",
]);

export class DJRequestAgent {
  constructor(
    private readonly llm: LLMRouter,
    private readonly llmTimeoutMs = 16000,
  ) {}

  async decide(userMessage: string, contextPack: MemoryPack): Promise<DJDecision> {
    const clean = compactText(userMessage, 500);
    if (!clean) return this.safeQuestion(clean, "Empty request.");

    const response = await this.chatJson(this.prompt(clean, contextPack)).catch(() => "");
    const data = extractJsonObject(response);
    if (!Object.keys(data).length) {
      return this.localSceneStyleDecision(clean) || this.safeQuestion(clean, "The DJ request agent did not return valid JSON.");
    }
    return this.decisionFromData(clean, data);
  }

  private async chatJson(prompt: string): Promise<string> {
    return this.llm.chat(prompt, {
      maxTokens: 650,
      system:
        "You are FREQME's private AI DJ request agent. Understand the listener's music intent from conversation context and return only valid JSON. Do not keyword-match the request.",
      responseFormat: { type: "json_object" },
      timeoutMs: this.llmTimeoutMs,
    });
  }

  private prompt(userMessage: string, contextPack: MemoryPack): string {
    return `The listener just spoke to an AI radio DJ:
${userMessage}

Compact memory/context JSON:
${JSON.stringify(this.promptContext(contextPack))}

Think like a conversational AI DJ with memory. Infer the real music intent, then return one structured decision.

Rules:
- Do not classify by local keywords. Reason from the whole utterance, session memory, recent turns, current track, profile, and constraints.
- Never ask downstream search to search the literal user sentence unless the sentence itself is already a clean canonical song/work title.
- Fuzzy artists, bands, performers, composers, works, scenes, genres, corrections, refusals, and continuation requests must become structured music tasks.
- A genre/style request is actionable by itself. R&B, jazz, hip-hop, city pop, shoegaze, Cantonese pop, or "quiet night music" should become scene_genre_direction, not a clarification.
- Most requests should play or set a direction; ask only when context still cannot resolve the ambiguity.
- Search goals are helpful but not mandatory. If the listener asks for a performer, composer, arranger, producer, style, scene, memory-based mood, or other abstract music direction, preserve that intent as structured entities and let the downstream search-planning agent infer concrete playable tracks.
- Only ask a clarification when the musical intent itself is genuinely ambiguous, not just because you cannot name a specific song yet.
- For negative feedback, capture immediate constraints without turning one skip into a permanent dislike.
- The DJ response should be short, natural Chinese. Do not mention systems, agents, algorithms, prompts, JSON, search, or models.

Return only JSON:
{
  "action": "play_now | set_direction_and_play | revise_mode_and_play | soft_confirm_and_play | ask_clarifying_question | negative_feedback | continue_current_mode",
  "understood_intent": "internal understanding",
  "music_task": {
    "type": "specific_track | artist_direction | artist_work_direction | scene_genre_direction | negative_feedback | continuation",
    "primary_entities": [{"role": "artist|performer|composer|arranger|producer|work|genre|scene|music_entity", "name": "canonical or inferred name"}],
    "work_hint": "",
    "style_hint": "",
    "negative_constraints": ["directions to avoid right now"],
    "search_goals": ["concrete query, preferably artist title"],
    "must_not_search_literal_user_sentence": true
  },
  "queue_policy": {"duration_tracks": 1, "continue_direction": false, "avoid_repetition": true},
  "uncertainty": {"level": "low|medium|high", "reason": "", "should_ask_user": false},
  "dj_response": {"speak_now": "short natural Chinese DJ reply", "tone": "warm_confident"},
  "memory_update": {"session_preference": [], "possible_long_term_preference": [], "negative_constraints": []}
}`;
  }

  private decisionFromData(rawText: string, data: Record<string, unknown>): DJDecision {
    const action = allowedActions.has(data.action as DJDecision["action"])
      ? (data.action as DJDecision["action"])
      : "ask_clarifying_question";
    const musicTask = this.normalizeMusicTask(this.field(data, "music_task", "musicTask"));
    if (playableActions.has(action) && !this.isExecutable(musicTask)) {
      return this.safeQuestion(rawText, "The agent returned a playable action without an executable music task.");
    }
    const queuePolicy = this.strengthenQueuePolicy(this.normalizeQueuePolicy(this.field(data, "queue_policy", "queuePolicy"), action), action, musicTask);
    return {
      action,
      understoodIntent: compactText(this.field(data, "understood_intent", "understoodIntent") || "", 300),
      musicTask,
      queuePolicy,
      uncertainty: this.normalizeUncertainty(data.uncertainty, action),
      djResponse: this.normalizeDjResponse(this.field(data, "dj_response", "djResponse")),
      memoryUpdate: this.normalizeMemoryUpdate(this.field(data, "memory_update", "memoryUpdate")),
      rawText,
    };
  }

  private safeQuestion(rawText: string, reason = ""): DJDecision {
    return {
      action: "ask_clarifying_question",
      understoodIntent: "The request was not clear enough to safely choose music.",
      musicTask: {
        type: "unclear",
        primaryEntities: [],
        workHint: "",
        styleHint: "",
        negativeConstraints: [],
        searchGoals: [],
        mustNotSearchLiteralUserSentence: true,
      },
      queuePolicy: { durationTracks: 0, continueDirection: false, avoidRepetition: true },
      uncertainty: { level: "high", reason, shouldAskUser: true },
      djResponse: { speakNow: "这句我没接稳，你是想听某个歌手，还是这种氛围？", tone: "warm_clarifying" },
      memoryUpdate: { sessionPreference: [], possibleLongTermPreference: [], negativeConstraints: [] },
      rawText,
    };
  }

  private localSceneStyleDecision(rawText: string): DJDecision | null {
    const normalized = rawText.toLowerCase();
    const styleMarkers: Array<{ test: RegExp; label: string; goals: string[] }> = [
      {
        test: /\bemo\b|情绪|忧郁|难过|emo 的歌|emo的歌/iu,
        label: "late-night emo / sad alt-pop",
        goals: ["late night emo", "sad indie emo", "sad alt-pop"],
      },
      {
        test: /\br\s*&?\s*b\b|\brnb\b/iu,
        label: "R&B",
        goals: ["warm R&B", "late-night R&B", "alt R&B"],
      },
      {
        test: /jazz|爵士/iu,
        label: "jazz",
        goals: ["vocal jazz", "late-night jazz", "soft jazz"],
      },
      {
        test: /ambient|氛围|放松|舒缓/iu,
        label: "ambient / chillout",
        goals: ["ambient", "chillout", "organic ambient"],
      },
      {
        test: /city\s*pop|citypop|城市流行/iu,
        label: "city pop",
        goals: ["city pop", "Japanese city pop"],
      },
      {
        test: /shoegaze|盯鞋/iu,
        label: "shoegaze",
        goals: ["shoegaze", "dream pop shoegaze"],
      },
    ];
    const marker = styleMarkers.find((item) => item.test.test(normalized));
    if (!marker) return null;
    const scenes = [
      /\bnight\b|晚上|夜晚|深夜|睡前/iu.test(rawText) ? "夜晚" : "",
      /雨|下雨|rain/iu.test(rawText) ? "雨天" : "",
      /开车|路上|通勤/iu.test(rawText) ? "路上" : "",
    ].filter(Boolean);
    const negativeConstraints = this.localNegativeConstraints(rawText);
    const styleHint = compactText([...scenes, marker.label].join(" "), 120);
    return {
      action: "set_direction_and_play",
      understoodIntent: compactText(`用户想听${styleHint || marker.label}，并希望电台持续沿着这个方向。`, 260),
      musicTask: {
        type: "scene_genre_direction",
        primaryEntities: [
          ...scenes.map((scene) => ({ role: "scene" as const, name: scene })),
          { role: "genre", name: marker.label },
        ],
        workHint: "",
        styleHint,
        negativeConstraints,
        searchGoals: marker.goals,
        mustNotSearchLiteralUserSentence: true,
      },
      queuePolicy: { durationTracks: 5, continueDirection: true, avoidRepetition: true },
      uncertainty: { level: "medium", reason: "LLM fallback for a clear scene/style request.", shouldAskUser: false },
      djResponse: { speakNow: "懂了，我先按这个氛围接住，后面也不乱跳。", tone: "warm_confident" },
      memoryUpdate: {
        sessionPreference: [styleHint || marker.label],
        possibleLongTermPreference: [],
        negativeConstraints,
      },
      rawText,
    };
  }

  private localNegativeConstraints(rawText: string): string[] {
    const constraints: string[] = [];
    const text = rawText.toLowerCase();
    if (/\bedm\b|电子舞曲|电音/iu.test(text)) constraints.push("EDM");
    if (/dubstep|回响贝斯/iu.test(text)) constraints.push("dubstep");
    if (/不要太炸|别太炸|不炸|太炸|炸场|高能/iu.test(rawText)) constraints.push("高能量");
    if (/不要中文|别中文|中文歌/iu.test(rawText)) constraints.push("中文歌");
    return constraints;
  }

  private normalizeMusicTask(value: unknown): MusicTask {
    const source = value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : {};
    const rawType = this.normalizeTaskType(compactText(source.type || "unclear", 40));
    const rawEntities = this.field(source, "primary_entities", "primaryEntities");
    const entities = Array.isArray(rawEntities)
      ? rawEntities
          .filter((item): item is Record<string, unknown> => Boolean(item && typeof item === "object" && !Array.isArray(item)))
          .map((item) => ({
            role: compactText(item.role || "music_entity", 40) as MusicTask["primaryEntities"][number]["role"],
            name: compactText(item.name || "", 120),
          }))
          .filter((item) => item.name)
          .slice(0, 5)
      : [];
    return {
      type: validTaskTypes.has(rawType) ? rawType : "unclear",
      primaryEntities: entities,
      workHint: compactText(this.field(source, "work_hint", "workHint") || "", 160),
      styleHint: compactText(this.field(source, "style_hint", "styleHint") || "", 180),
      negativeConstraints: asStringList(this.field(source, "negative_constraints", "negativeConstraints"), 10),
      searchGoals: asStringList(this.field(source, "search_goals", "searchGoals"), 8),
      mustNotSearchLiteralUserSentence: true,
    };
  }

  private normalizeQueuePolicy(value: unknown, action: DJDecision["action"]): DJDecision["queuePolicy"] {
    const source = value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : {};
    const duration = Number(this.field(source, "duration_tracks", "durationTracks") ?? (action === "play_now" ? 1 : 4));
    const continueDirection = this.field(source, "continue_direction", "continueDirection");
    const avoidRepetition = this.field(source, "avoid_repetition", "avoidRepetition");
    return {
      durationTracks: Math.max(0, Math.min(8, Number.isFinite(duration) ? Math.floor(duration) : 1)),
      continueDirection: typeof continueDirection === "boolean" ? continueDirection : action !== "play_now",
      avoidRepetition: typeof avoidRepetition === "boolean" ? avoidRepetition : true,
    };
  }

  private strengthenQueuePolicy(
    policy: DJDecision["queuePolicy"],
    action: DJDecision["action"],
    task: MusicTask,
  ): DJDecision["queuePolicy"] {
    if (
      ["set_direction_and_play", "revise_mode_and_play", "continue_current_mode", "negative_feedback"].includes(action) &&
      ["artist_direction", "artist_work_direction", "scene_genre_direction", "continuation", "negative_feedback"].includes(task.type)
    ) {
      return {
        ...policy,
        durationTracks: Math.max(policy.durationTracks, 4),
        continueDirection: true,
      };
    }
    return policy;
  }

  private normalizeUncertainty(value: unknown, action: DJDecision["action"]): DJDecision["uncertainty"] {
    const source = value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : {};
    const level = source.level === "low" || source.level === "high" ? source.level : "medium";
    const shouldAskUser = this.field(source, "should_ask_user", "shouldAskUser");
    return {
      level,
      reason: compactText(source.reason || "", 160),
      shouldAskUser: typeof shouldAskUser === "boolean" ? shouldAskUser : action === "ask_clarifying_question",
    };
  }

  private normalizeDjResponse(value: unknown): DJDecision["djResponse"] {
    const source = value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : {};
    return {
      speakNow: compactText(this.field(source, "speak_now", "speakNow") || "我先按我理解到的方向接上。", 180),
      tone: compactText(source.tone || "warm_confident", 60),
    };
  }

  private normalizeMemoryUpdate(value: unknown): DJDecision["memoryUpdate"] {
    const source = value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : {};
    return {
      sessionPreference: asStringList(this.field(source, "session_preference", "sessionPreference"), 8),
      possibleLongTermPreference: asStringList(this.field(source, "possible_long_term_preference", "possibleLongTermPreference"), 8),
      negativeConstraints: asStringList(this.field(source, "negative_constraints", "negativeConstraints"), 8),
    };
  }

  private isExecutable(task: MusicTask): boolean {
    return Boolean(
      task.searchGoals.length ||
        task.primaryEntities.length ||
        task.workHint.trim() ||
        task.styleHint.trim() ||
        task.negativeConstraints.length,
    );
  }

  private promptContext(contextPack: MemoryPack): Record<string, unknown> {
    const playback = contextPack.playbackContext || {};
    const settings = contextPack.userSettings || {};
    return {
      profile: compactText(contextPack.userProfileDigest, 650),
      session: this.compactJson(contextPack.sessionWorkingMemory, 650),
      recentTurns: contextPack.recentTurns.slice(-4),
      memories: contextPack.retrievedMemories.slice(0, 4),
      playback: {
        currentTrack: (playback as Record<string, unknown>).currentTrack,
        recentTracks: Array.isArray((playback as Record<string, unknown>).recentTracks)
          ? ((playback as Record<string, unknown>).recentTracks as unknown[]).slice(-5)
          : [],
        scene: (playback as Record<string, unknown>).scene,
      },
      settings: {
        currentMode: settings.currentMode,
        musicNotes: compactText(settings.musicNotes || "", 220),
        timezoneName: settings.timezoneName,
        locale: settings.locale,
        regionHint: settings.regionHint,
        localTimeBlock: settings.localTimeBlock,
      },
      constraints: contextPack.hardConstraints.slice(0, 5),
    };
  }

  private compactJson(value: unknown, maxLength: number): string {
    return compactText(JSON.stringify(value ?? {}), maxLength);
  }

  private field(source: Record<string, unknown>, snake: string, camel: string): unknown {
    return source[snake] ?? source[camel];
  }

  private normalizeTaskType(value: string): MusicTask["type"] {
    const aliases: Record<string, MusicTask["type"]> = {
      genre_direction: "scene_genre_direction",
      style_direction: "scene_genre_direction",
      mood_direction: "scene_genre_direction",
      scene_direction: "scene_genre_direction",
      artist: "artist_direction",
      track: "specific_track",
      song: "specific_track",
    };
    return (aliases[value] || value) as MusicTask["type"];
  }
}
